NativeImages are slow to create from remote contexts

kkinnunen-apple opened the pull request:
#### 827b5833cabfb4d2444d2c4da95c8489fd4fcdc3
<pre>
NativeImages are slow to create from remote contexts
<a href="https://bugs.webkit.org/show_bug.cgi?id=266215">https://bugs.webkit.org/show_bug.cgi?id=266215</a>
<a href="https://rdar.apple.com/119489233">rdar://119489233</a>

Reviewed by Matt Woodrow.

RemoteImageBufferProxy::copyNativeImage() would transfer the rendering
from GPUP to WCP. This was slow, and in turn lead to introduction of
ImageBuffer as a source for all types of bitmap draws, since this
resource exist in GPUP. This in turn is problematic, as ImageBuffers
are inherently mutable and as such cannot be shared. ImageBuffer is
a draw target type of abstraction, where as NativeImage is draw source
type of abstraction.

Fix this by implementing RemoteImageBufferProxy::copyNativeImage()
as a command that creates the NativeImage in GPUP and returns a
RemoteNativeImageProxy for WCP code. The image is
constructed in GPUP and can be referred to in all the draw operations
taking NativeImage. If the image data is used locally, the
data is synchronized from GPUP to WCP at that point.

Since RemoteNativeImageProxyClient is using WeakPtrs similar to existing
RenderingResourceObserver, convert RemoteResourceCacheProxy as a
heap allocated type that can be used with CheckedPtr. Fix the tech
debt in RenderingResourceObserver, and make it CheckedPtr type instead
of opting in for smart pointer rule exception.

This commit introduces a edge-case regression where a GPUP process
crash will lose the GPUP-side NativeImage data in case it was not
explicitly used in WCP before. Currently there are no call sites that
form long-term references. Later commits will introduce best-effort
synchronization where the long-term held NativeImage data is
sent from GPUP to WCP.

Not many call sites use NativeImages. ImageBitmap is one such case.
This speeds up PerformanceTests/Canvas/drawImageSourceImageBitmapCanvas2D.html
from 850ms to 650ms, but future commits will improve this substantially
more.

This is first commit in a series that will convert all image source
holders to use NativeImage, improving performance and resource use.
In future commits, NativeImage will be also made thread-safe, so that
instances references can be shared as well as passed freely across
threads.

* Source/WebCore/platform/graphics/Gradient.cpp:
(WebCore::Gradient::~Gradient):
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::~NativeImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/RenderingResource.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
* Source/WebCore/platform/graphics/displaylists/DisplayList.cpp:
(WebCore::DisplayList::DisplayList::~DisplayList):
* Source/WebCore/platform/graphics/filters/FilterFunction.cpp:
(WebCore::FilterFunction::~FilterFunction):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::copyNativeImage):
(WebKit::RemoteImageBuffer::getShareableBitmap): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::nativeImageBitmap):
(WebKit::RemoteRenderingBackend::cacheNativeImage):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheNativeImage):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::send const):
(WebKit::RemoteImageBufferProxy::sendSync const):
(WebKit::RemoteImageBufferProxy::copyNativeImage const):
(WebKit::RemoteImageBufferProxy::send): Deleted.
(WebKit::RemoteImageBufferProxy::sendSync): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageProxy.cpp: Added.
(WebKit::placeholderPlatformImage):
(WebKit::RemoteNativeImageProxy::create):
(WebKit::RemoteNativeImageProxy::RemoteNativeImageProxy):
(WebKit::RemoteNativeImageProxy::~RemoteNativeImageProxy):
(WebKit::RemoteNativeImageProxy::platformImage const):
(WebKit::RemoteNativeImageProxy::size const):
(WebKit::RemoteNativeImageProxy::hasAlpha const):
(WebKit::RemoteNativeImageProxy::colorSpace const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::nativeImageCountForTesting const):
(WebKit::RemoteRenderingBackendProxy::createNativeImage):
(WebKit::RemoteRenderingBackendProxy::nativeImageBitmap):
(WebKit::RemoteRenderingBackendProxy::releaseMemory):
(WebKit::RemoteRenderingBackendProxy::releaseNativeImages):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
(WebKit::RemoteRenderingBackendProxy::getShareableBitmap): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::remoteResourceCacheProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::create):
(WebKit::RemoteResourceCacheProxy::createNativeImage):
(WebKit::RemoteResourceCacheProxy::willDestroyRemoteNativeImageProxy):
(WebKit::RemoteResourceCacheProxy::platformImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: <a href="https://commits.webkit.org/300811@main">https://commits.webkit.org/300811@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97e7b8829fc91c6cf79fd3721520011c67a229c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34124 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130488 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75859 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b9f5d3ce-7880-4de5-816d-62c74e38e0f6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125590 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52022 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94075 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62438 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/010dcea5-8a13-47b9-8e3c-c3a8dbd8c9ce) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126666 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110672 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74680 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7833dad9-15df-409d-9bef-769dbc033511) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34130 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28832 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73970 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29055 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133177 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38573 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102549 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51039 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102386 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26088 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47730 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25978 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47576 "Hash 97e7b882 for PR 51251 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50518 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56280 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49993 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53339 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51667 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/platform/graphics/RenderingResource.h:
```diff
@@ -49,6 +50,7 @@ class RenderingResourceObserver : public CanMakeWeakPtr<RenderingResourceObserve
 public:
     virtual ~RenderingResourceObserver() = default;
     virtual void willDestroyNativeImage(const NativeImage&) = 0;
+    virtual PlatformImagePtr platformImageForNativeImage(const NativeImage&) = 0;
```
This feels a bit awkward. It's not obviously the role of the 'observer' to be the one to do the lookup of a platform image, and needing to iterate an arbitrary set of observers to see if any of them are willing to offer a platform image feels wrong.

I assume you considered a `RemoteNativeImageProxy` subclass instead?

It still feels like `NativeImage` is trying to simultaneously be the 'immutable snapshot (hopefully copy-on-write underneath) of an ImageBuffer, used as a draw source', and also the 'native' bit where the caller actually wants to get at a platform object.

kkinnunen-apple replied:
>I assume you considered a RemoteNativeImageProxy subclass instead?

Yeah, it's the same structure, though, nothing substantially different there.

 The issue with that is that either:
a) There is a NativeImage interface and NativeImagePlatform, RemoteNativeImageProxy implementations. This is problematic because the change is larger and people have hard time adjusting their mental model wrt NativeImagePlatform 
or
b) The subclass relationship is not correct: RemoteNativeImageProxy is-a NativeImage does not really hold if NativeImage is a concrete class and not an interface. In other words, the inheritance here is used as a mechanism instead of signifying subtyping.

The updated patch contains the b) since it's simpler to get acceptance.

> It still feels like NativeImage is trying to simultaneously be the 'immutable snapshot (hopefully copy-on-write underneath) of an ImageBuffer, used as a draw source', and also the 'native' bit where the caller actually wants to get at a platform object.

The former is a correct model of thought. NativeImage (or CGImage or SkImage) is a representation of drawing or loaded image.

The latter is unsubstantiated. There's no real use-case for NativeImage to be " 'native' bit where the caller actually wants to get at a platform object".

Image is something one can get the data out of.  The platform object is the data.

mattwoodrow commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -343,20 +343,51 @@ void RemoteRenderingBackend::destroyGetPixelBufferSharedMemory()
         return;
 
-    auto image = NativeImage::create(bitmap->createPlatformImage(DontCopyBackingStore, ShouldInterpolate::Yes), nativeImageIdentifier);
+    auto image = NativeImage::create(bitmap->createPlatformImage(DontCopyBackingStore, ShouldInterpolate::Yes));
```
Why did you remove the passing of the identifier here?

kkinnunen-apple replied:
The patch omits the identifier selection in RemoteNativeImageProxy, because that is the "correct" approach. As such, it is consistent to omit it in NativeImage. So the identifier selection was simplified to be just a ::generate() on the NativeImage::m_identifier.

The pass of the identifier here was just a convention: the GPUP side NativeImage instances would get the same identifier as which it was cached in its particular RRB cache. However, the identifier was not really used anywhere in GPUP side. 

So the caching was changed:
- From: "Cache this NativeImage to the identifier it holds"
- To: "Cache this NativeImage to this identifier"

This establishes the code in direction of "normal" resources: the resources themselves do not have identifiers, rather the identifier is just specific to the use cases in the IPC communication.

In here it's only done half-away, e.g. the code is partially adjusted. NativeImages still hold identifiers, since SourceImage accesses it directly and incorrectly. Later on, will be fixed and the line above is already done.

mattwoodrow replied:
> In here it's only done half-away, e.g. the code is partially adjusted. NativeImages still hold identifiers, since SourceImage accesses it directly and incorrectly. Later on, will be fixed and the line above is already done.

Right, of course, I should have remembered that you'd already done this for other types.

It seemed a bit strange for the identifier keying the hash map to not match the identifier of the object, even though it was unused. Makes sense in the context of the larger plan though!

kkinnunen-apple committed:
NativeImages are slow to create from remote contexts
https://bugs.webkit.org/show_bug.cgi?id=266215
rdar://119489233

Reviewed by Matt Woodrow.

RemoteImageBufferProxy::copyNativeImage() would transfer the rendering
from GPUP to WCP. This was slow, and in turn lead to introduction of
ImageBuffer as a source for all types of bitmap draws, since this
resource exist in GPUP. This in turn is problematic, as ImageBuffers
are inherently mutable and as such cannot be shared. ImageBuffer is
a draw target type of abstraction, where as NativeImage is draw source
type of abstraction.

Fix this by implementing RemoteImageBufferProxy::copyNativeImage()
as a command that creates the NativeImage in GPUP and returns a
RemoteNativeImageProxy for WCP code. The image is
constructed in GPUP and can be referred to in all the draw operations
taking NativeImage. If the image data is used locally, the
data is synchronized from GPUP to WCP at that point.

Since RemoteNativeImageProxyClient is using WeakPtrs similar to existing
RenderingResourceObserver, convert RemoteResourceCacheProxy as a
heap allocated type that can be used with CheckedPtr. Fix the tech
debt in RenderingResourceObserver, and make it CheckedPtr type instead
of opting in for smart pointer rule exception.

This commit introduces a edge-case regression where a GPUP process
crash will lose the GPUP-side NativeImage data in case it was not
explicitly used in WCP before. Currently there are no call sites that
form long-term references. Later commits will introduce best-effort
synchronization where the long-term held NativeImage data is
sent from GPUP to WCP.

Not many call sites use NativeImages. ImageBitmap is one such case.
This speeds up PerformanceTests/Canvas/drawImageSourceImageBitmapCanvas2D.html
from 850ms to 650ms, but future commits will improve this substantially
more.

This is first commit in a series that will convert all image source
holders to use NativeImage, improving performance and resource use.
In future commits, NativeImage will be also made thread-safe, so that
instances references can be shared as well as passed freely across
threads.

* Source/WebCore/platform/graphics/Gradient.cpp:
(WebCore::Gradient::~Gradient):
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::~NativeImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/RenderingResource.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::createTransient):
* Source/WebCore/platform/graphics/displaylists/DisplayList.cpp:
(WebCore::DisplayList::DisplayList::~DisplayList):
* Source/WebCore/platform/graphics/filters/FilterFunction.cpp:
(WebCore::FilterFunction::~FilterFunction):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::copyNativeImage):
(WebKit::RemoteImageBuffer::getShareableBitmap): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::nativeImageBitmap):
(WebKit::RemoteRenderingBackend::cacheNativeImage):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheNativeImage):
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::send const):
(WebKit::RemoteImageBufferProxy::sendSync const):
(WebKit::RemoteImageBufferProxy::copyNativeImage const):
(WebKit::RemoteImageBufferProxy::send): Deleted.
(WebKit::RemoteImageBufferProxy::sendSync): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageProxy.cpp: Added.
(WebKit::placeholderPlatformImage):
(WebKit::RemoteNativeImageProxy::create):
(WebKit::RemoteNativeImageProxy::RemoteNativeImageProxy):
(WebKit::RemoteNativeImageProxy::~RemoteNativeImageProxy):
(WebKit::RemoteNativeImageProxy::platformImage const):
(WebKit::RemoteNativeImageProxy::size const):
(WebKit::RemoteNativeImageProxy::hasAlpha const):
(WebKit::RemoteNativeImageProxy::colorSpace const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageProxy.h: Added.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::nativeImageCountForTesting const):
(WebKit::RemoteRenderingBackendProxy::createNativeImage):
(WebKit::RemoteRenderingBackendProxy::nativeImageBitmap):
(WebKit::RemoteRenderingBackendProxy::releaseMemory):
(WebKit::RemoteRenderingBackendProxy::releaseNativeImages):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
(WebKit::RemoteRenderingBackendProxy::getShareableBitmap): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::remoteResourceCacheProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::create):
(WebKit::RemoteResourceCacheProxy::createNativeImage):
(WebKit::RemoteResourceCacheProxy::willDestroyRemoteNativeImageProxy):
(WebKit::RemoteResourceCacheProxy::platformImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:

Canonical link: https://commits.webkit.org/300811@main
